### PR TITLE
Common refactoring part 1/X

### PR DESCRIPTION
**Pull request by wakr**
Mainly checkstyle fixing and removing unused methods. Introduced a Factory-pattern for Command-classes. This factory is mainly used atm by TmcCore. This reduced the amount of direct dependencies which TmcCore had alot.

### CODE REVIEW

File: src/main/java/fi/helsinki/cs/tmc/core/TmcCore.java [hunk 1]
```diff
@@ -1,29 +1,11 @@
 package fi.helsinki.cs.tmc.core;
 
-import static com.google.common.base.Strings.isNullOrEmpty;
+import static fi.helsinki.cs.tmc.core.commands.factory.CommandFactory.*;
```
Comment (jamo) [inline src/main/java/fi/helsinki/cs/tmc/core/TmcCore.java]: Sorry, no .\* imports: https://google.github.io/styleguide/javaguide.html#s3.3.1-wildcard-imports
Event though it make the import list longer..

File: src/main/java/fi/helsinki/cs/tmc/core/TmcCore.java [hunk 2]
```diff
@@ -203,7 +162,7 @@ public Path getExerciseChecksumCacheLocation() {
      * @param path inside any exercise directory
      * @return SubmissionResult object containing details of the tests run on server
      * @throws TmcCoreException if there was no course in the given path, no exercise in the given
-     * path, or not logged in
+     path, or not logged in.
```
Comment (jamo) [inline src/main/java/fi/helsinki/cs/tmc/core/TmcCore.java]: indentation is incorrect:  When an at-clause doesn't fit on a single line, continuation lines are indented four (or more) spaces from the position of the @.
https://google.github.io/styleguide/javaguide.html#s7.1.3-javadoc-at-clauses

File: src/main/java/fi/helsinki/cs/tmc/core/TmcCore.java [hunk 3]
```diff
@@ -346,9 +286,9 @@ public Path getExerciseChecksumCacheLocation() {
      */
     public ListenableFuture<URI> requestCodeReview(Path path, String message)
             throws TmcCoreException {
-        //checkParameters(path);
-        RequestCodeReview request = new RequestCodeReview(settings, path, message);
-        return threadPool.submit(request);
+
+        Command<URI> requestCodeReview = getRequestCodeReviewCmd(settings, path, message);
+        return threadPool.submit(requestCodeReview);
```
Comment (jamo) [inline src/main/java/fi/helsinki/cs/tmc/core/TmcCore.java]: I ques we could inline this.

File: src/main/java/fi/helsinki/cs/tmc/core/commands/SendSpywareDiffs.java [hunk 4]
```diff
@@ -30,22 +31,34 @@ public SendSpywareDiffs(TmcSettings settings, DiffSender sender, byte[] spywareD
...
-        Optional<Course> course = this.settings.getCurrentCourse();
-        if (course == null || !course.isPresent()) {
-            throw new TmcCoreException("No current course found from settings.");
+    private void testUsername() throws TmcCoreException {
+        if (isBadString(settings.getUsername())) {
+            throw new TmcCoreException("username must be set!");
         }
     }
 
+    private boolean isBadString(String toBeTested) {
```
Comment (jamo) [inline src/main/java/fi/helsinki/cs/tmc/core/commands/SendSpywareDiffs.java]: Not really sure if this method really is worth it... Maybe static import the isNullOrEmpty...

File: src/main/java/fi/helsinki/cs/tmc/core/commands/Submit.java [hunk 5]
```diff
@@ -56,17 +57,26 @@ public Submit(
...
         }
     }
 
+    private void checkUsername() throws TmcCoreException {
+        if (isBadString(settings.getUsername())) {
+            throw new TmcCoreException("username must be set!");
+        }
+    }
+
+    private boolean isBadString(String toBeTested) {
```
Comment (jamo) [inline src/main/java/fi/helsinki/cs/tmc/core/commands/Submit.java]: Yeah, lets just use the staticly imported `isNullOrEmpty` :)

File: src/main/java/fi/helsinki/cs/tmc/core/commands/factory/CommandFactory.java [hunk 6]
```diff
@@ -0,0 +1,223 @@
+package fi.helsinki.cs.tmc.core.commands.factory;
+
+import fi.helsinki.cs.tmc.core.cache.ExerciseChecksumCache;
+import fi.helsinki.cs.tmc.core.commands.*;
+import fi.helsinki.cs.tmc.core.communication.*;
```
Comment (jamo) [inline src/main/java/fi/helsinki/cs/tmc/core/commands/factory/CommandFactory.java]: No `.*` imports

File: src/main/java/fi/helsinki/cs/tmc/core/commands/factory/CommandFactory.java [hunk 7]
```diff
@@ -0,0 +1,223 @@
...
+
+/**
+ * Handles the creation of different Commands. Mainly used by
+ * {@link fi.helsinki.cs.tmc.core.TmcCore}
+ */
+
+public class CommandFactory {
+
+    /**
+     * .
```
Comment (jamo) [inline src/main/java/fi/helsinki/cs/tmc/core/commands/factory/CommandFactory.java]: Don't really like how these JavaDocs are -- we should also avoid just having at-clauses w/o real description

File: src/main/java/fi/helsinki/cs/tmc/core/util/ParameterTester.java [hunk 8]
```diff
@@ -0,0 +1,20 @@
...
+    public static void checkStringParameters(String...params) throws TmcCoreException {
+        for (String param : params) {
+            if (isNullOrEmpty(param)) {
+                throw new TmcCoreException("Param {" + param + "}empty or null.");
+            }
+        }
+    }
+
+
+
```
Comment (jamo) [inline src/main/java/fi/helsinki/cs/tmc/core/util/ParameterTester.java]: too many blank lines

**Comment by jamo** [comment]
Nice work!
Added a few comments and nitpicks.

File: src/main/java/fi/helsinki/cs/tmc/core/TmcCore.java [hunk 9]
```diff
@@ -147,41 +126,21 @@ public Path getExerciseChecksumCacheLocation() {
...
-     */
-    public ListenableFuture<List<Exercise>> downloadExercises(List<Exercise> exercises)
-            throws TmcCoreException {
-        return this.downloadExercises(exercises, null);
-    }
-
-    /**
-     * Downloads exercises.
-     */
-    public ListenableFuture<List<Exercise>> downloadExercises(
```
Comment (ljleppan) [inline src/main/java/fi/helsinki/cs/tmc/core/TmcCore.java:174]: I'm unsure about whether we should remove this outright. A more cautious approach would be to `@deprecate` it first.

File: src/main/java/fi/helsinki/cs/tmc/core/TmcCore.java (hunk 2 above)
Comment (ljleppan) [inline src/main/java/fi/helsinki/cs/tmc/core/TmcCore.java]: Note that this same issue affects multiple places in this PR.

File: src/main/java/fi/helsinki/cs/tmc/core/commands/SendSpywareDiffs.java (hunk 4 above)
Comment (ljleppan) [inline src/main/java/fi/helsinki/cs/tmc/core/commands/SendSpywareDiffs.java]: Also, "bad" is a bit unclear. +1 to replacing with statically imported `isNullOrEmpty(String string)`.

File: src/main/java/fi/helsinki/cs/tmc/core/commands/Submit.java [hunk 10]
```diff
@@ -1,5 +1,6 @@
 package fi.helsinki.cs.tmc.core.commands;
 
+import com.google.common.base.Strings;
```
Comment (ljleppan) [inline src/main/java/fi/helsinki/cs/tmc/core/commands/Submit.java]: Import order

File: src/main/java/fi/helsinki/cs/tmc/core/commands/VerifyCredentials.java [hunk 11]
```diff
@@ -1,5 +1,6 @@
 package fi.helsinki.cs.tmc.core.commands;
 
+import com.google.common.base.Strings;
```
Comment (ljleppan) [inline src/main/java/fi/helsinki/cs/tmc/core/commands/VerifyCredentials.java]: Import order

File: src/main/java/fi/helsinki/cs/tmc/core/commands/VerifyCredentials.java [hunk 12]
```diff
@@ -52,6 +62,10 @@ public Boolean call() throws TmcCoreException, IOException {
                 .makeGetRequest(URI.create(settings.getServerAddress() + TMC_SERVER_ROUTE), auth)
                 .getStatusCode();
 
-        return (response >= HTTP_SUCCESS_RANGE_MIN && response <= HTTP_SUCCESS_RANGE_MAX);
+        return isResponseInRange(response);
+    }
+
+    private boolean isResponseInRange(int response) {
```
Comment (ljleppan) [inline src/main/java/fi/helsinki/cs/tmc/core/commands/VerifyCredentials.java]: Consider renaming to `isStatusCodeSuccess()` or something similar, "in range" leaves much for interpretation.

File: src/main/java/fi/helsinki/cs/tmc/core/TmcCore.java [hunk 13]
```diff
@@ -147,41 +138,21 @@ public Path getExerciseChecksumCacheLocation() {
...
-        return threadPool.submit(downloadCommand);
-    }
-
-    /**
-     * Downloads exercise files specified in the given list. The exercises will be located in
-     * TmcMainDirectory (field in TmcSettings).
-     *
-     * @param exercises to be downloaded
-     */
-    public ListenableFuture<List<Exercise>> downloadExercises(List<Exercise> exercises)
```
Comment (ljleppan) [inline src/main/java/fi/helsinki/cs/tmc/core/TmcCore.java:166]: I'm unsure about whether we should remove this outright. A more cautious approach would be to `@deprecate` it first.

File: src/main/java/fi/helsinki/cs/tmc/core/commands/factory/CommandFactory.java [hunk 14]
```diff
@@ -0,0 +1,168 @@
+package fi.helsinki.cs.tmc.core.commands.factory;
+
+import fi.helsinki.cs.tmc.core.cache.ExerciseChecksumCache;
+
+
```
Comment (ljleppan) [inline src/main/java/fi/helsinki/cs/tmc/core/commands/factory/CommandFactory.java]: Nit: Extra lines

**Comment by ljleppan** [comment]
Nice work! Added a bunch comments. Mostly nit-picky stuff about imports etc., but also few about method naming.

**Comment by jamo** [comment]
Nice work!
Merged with a few nitpicks I made directly.

**Pull request merged**